File: 17_Binary Tree-I/E094. Inorder Traversal.java
// 94. Binary Tree Inorder Traversal
// https://leetcode.com/problems/binary-tree-inorder-traversal/

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class Solution {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        
        while (root != null || !stack.isEmpty()) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
            root = stack.pop();
            result.add(root.val);
            root = root.right;
        }
        
        return result;
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();
        TreeNode root = new TreeNode(1);
        root.right = new TreeNode(2);
        root.right.left = new TreeNode(3);
        
        List<Integer> result = solution.inorderTraversal(root);
        System.out.println(result);
    }
}

File: JavaScript/17_Binary Tree-I/E094. Inorder Traversal.js
// 94. Binary Tree Inorder Traversal
// https://leetcode.com/problems/binary-tree-inorder-traversal/

/**
 * TreeNode class for binary tree
 */
/**
 * Creates a tree node
 */
function createTreeNode(val = 0, left = null, right = null) {
    return { val, left, right };
}

/**
 * Iterative Inorder Traversal using stack
 * @param {TreeNode} root - Root of the binary tree
 * @returns {number[]} - Inorder traversal (Left -> Root -> Right)
 */
function inorderTraversal(root) {
    const result = [];
    const stack = [];
    let curr = root;

    while (curr !== null || stack.length > 0) {
        // Go to leftmost node
        while (curr !== null) {
            stack.push(curr);
            curr = curr.left;
        }

        // Process node
        curr = stack.pop();
        result.push(curr.val);

        // Move to right subtree
        curr = curr.right;
    }

    return result;
}

/**
 * Recursive Inorder Traversal
 * @param {TreeNode} root - Root of the binary tree
 * @returns {number[]} - Inorder traversal
 */
function inorderTraversalRecursive(root) {
    const result = [];

    function inorder(node) {
        if (node === null) return;
        inorder(node.left);
        result.push(node.val);
        inorder(node.right);
    }

    inorder(root);
    return result;
}

// Test cases
const root = createTreeNode(1);
root.right = createTreeNode(2);
root.right.left = createTreeNode(3);

console.log(inorderTraversal(root)); // [1, 3, 2]
console.log(inorderTraversalRecursive(root)); // [1, 3, 2]